sys/linux/dev_msm.txt

```diff
@@ -0,0 +1,147 @@
+# Copyright 2021 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <drm/msm_drm.h>
+
+resource fd_msm[fd]
+resource fd_msm_fence[fd]
+
+openat$msm(fd const[AT_FDCWD], file ptr[in, string["/dev/msm"]], flags flags[open_flags], mode const[0]) fd_msm
+
+ioctl$DRM_IOCTL_MSM_GET_PARAM(fd fd_msm, cmd const[DRM_IOCTL_MSM_GET_PARAM], arg ptr[inout, drm_msm_param])
+ioctl$DRM_IOCTL_MSM_GEM_NEW(fd fd_msm, cmd const[DRM_IOCTL_MSM_GEM_NEW], arg ptr[inout, drm_msm_gem_new])
+ioctl$DRM_IOCTL_MSM_GEM_INFO(fd fd_msm, cmd const[DRM_IOCTL_MSM_GEM_INFO], arg ptr[inout, drm_msm_gem_info])
+ioctl$DRM_IOCTL_MSM_GEM_CPU_PREP(fd fd_msm, cmd const[DRM_IOCTL_MSM_GEM_CPU_PREP], arg ptr[inout, drm_msm_gem_cpu_prep])
+ioctl$DRM_IOCTL_MSM_GEM_CPU_FINI(fd fd_msm, cmd const[DRM_IOCTL_MSM_GEM_CPU_FINI], arg ptr[in, drm_msm_gem_cpu_fini])
+ioctl$DRM_IOCTL_MSM_GEM_SUBMIT(fd fd_msm, cmd const[DRM_IOCTL_MSM_GEM_SUBMIT], arg ptr[inout, drm_msm_gem_submit]) fd_msm_fence
+ioctl$DRM_IOCTL_MSM_WAIT_FENCE(fd fd_msm, cmd const[DRM_IOCTL_MSM_WAIT_FENCE], arg ptr[in, drm_msm_wait_fence])
+ioctl$DRM_IOCTL_MSM_GEM_MADVISE(fd fd_msm, cmd const[DRM_IOCTL_MSM_GEM_MADVISE], arg ptr[inout, drm_msm_gem_madvise])
+ioctl$DRM_IOCTL_MSM_SUBMITQUEUE_NEW(fd fd_msm, cmd const[DRM_IOCTL_MSM_SUBMITQUEUE_NEW], arg ptr[inout, drm_msm_submitqueue])
+ioctl$DRM_IOCTL_MSM_SUBMITQUEUE_CLOSE(fd fd_msm, cmd const[DRM_IOCTL_MSM_SUBMITQUEUE_CLOSE], arg ptr[in, int32])
+ioctl$DRM_IOCTL_MSM_SUBMITQUEUE_QUERY(fd fd_msm, cmd const[DRM_IOCTL_MSM_SUBMITQUEUE_QUERY], arg ptr[inout, drm_msm_submitqueue_query])
+
+mmap$DRM_MSM(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_msm, offset fileoff)
+_ = __NR_mmap2
+
+type msm_handle int32
```

Comment:
Should it also be a resource?
`type msm_handle int32 ` is just a `typedef` to int32, and says "pass random int32 values always". Which is usually not what "handle" means.

Reply:
Sorry I was copying the bifrost description which seem to use an int32.  Sometimes they're gem handles and other times they're syncobjs. What should I do here?

Reply:
Unions can be used to express "one or another".
Or sometimes it can make sense to create several specializations of a struct or syscall to also say "_when_ it's one or another".
For struct specializations you can see e.g. file_handle using templates (but there are dozens of other examples):
https://github.com/google/syzkaller/blob/c0ced5570c6505a80235b64884c93e89733b77f3/sys/linux/sys.txt#L810-L851
For syscalls see e.g. fcntl:
https://github.com/google/syzkaller/blob/c0ced5570c6505a80235b64884c93e89733b77f3/sys/linux/sys.txt#L206-L223

Reply:
Thanks for the background. I suspect it will be good to describe the kernel's DRM ioctls in include/uapi/drm/drm.h then and use those definitions of drm_syncobj_* structs and associated ioctls to provide a more specific handle here. I think sys/linux/dev_ion.txt is defining the dma buf fd ioctl so maybe after the drm ioctls are defined for the sync obj we can make a "one or the other" definition for the handle.

Reply:
It would be good to add as TODO in the descriptions.
It's hard to tell what's "done" versus "needs more improvement". There are also lots of people looking at these descriptions who has no idea about all these subsystems (e.g. me :)).

Reply:
Sure no problem.

Reply:
I see that gem handles are already described in sys/linux/dev_dri.txt and the syncobjs are there too, just commented out. I'll push an update here that replaces this msm_handle with gem or drm sync objects. Then I'll try to add the ioctls that create sync objects a little later this week to see if that can start sending in proper objects to these msm ioctls.